### ISSUE TITLE
"Never" should be assignable to any type

### ISSUE THREAD

**Issue by russelldavis** [microsoft/pyright#851]
Pyright's type narrowing is smart enough to account for possible mutation of global variables:

```python
from typing import Optional
opt1: Optional[int] = None
reveal_type(opt1) # None

def foo():
    reveal_type(opt1) # int | None
```

But it doesn't make the same considerations for `nonlocal` references, which results in types that are too narrow:

```python
from typing import Optional
def bar():
    opt2: Optional[int] = None
    reveal_type(opt2) # None
    def qux():
        nonlocal opt2
        opt2 = 1
    mystery(qux)
    reveal_type(opt2) # None

def mystery(callback):
    ...
```

Since `mystery` might call `qux`, `opt2` should retain its original un-narrowed type of `Optional[int]` (at least at any point after `qux` is called or passed to a function).

**Comment by russelldavis** [comment]
The same idea applies for type inference. In the example, if you remove the type annotation from `opt2`, I would still expect the second `reveal_type` to show `Optional[int]` (in the same way it would if you later assigned it an `int` inside a conditional).

**Comment by erictraut** [comment]
There are certain assumptions the type checker needs to make when analyzing types. One key assumption is that values of symbols are modified locally within an "execution context" (i.e. a module, function, or lambda) and not modified in unpredictable ways outside of an execution context, even in cases where it's theoretically possible. This includes globals, non-locals, symbols captured in closures, and instance variables.

Let's look at an example using instance variables:
```
def foo(self):
    self._var1 = 4
    a = self._var1
```

In this case it's generally safe to assume that the type of `a` is `Literal[4]`, but what if `_var1` is implemented with a property whose getter does something non-obvious like returning a string some of the time? Our assumption in this contrived case would fail, but it's a good assumption for almost all real-world code. If we can't make this assumption, then type narrowing and proper type inference become impossible in many cases, so it's a tradeoff that's worth making.

The same assumption applies for globals and non-locals. Your example above breaks the assumption made in the code flow engine. There's no way for a type checker to know generally that a function call is going to mutate a non-local or global and assign it a different type. It can't assume that a callee is going to call a callback synchronously (or at all).

One option is to assume that all function calls (and therefore all operations that can be overridden with magic methods in Python) mutate all symbols outside of the immediate scope. That would effectively render most type narrowing and inference impossible, so it would be a bad tradeoff. 

I will assert that cases where a non-local is being mutated (like the example you provided above) are not only rare but are generally bad coding practices and should be avoided. In your example, it would be much better to return a new value from the subroutine call rather than rely on a "spooky action at a distance".

I'll also point out that these assumptions that are built in to Pyright's code flow engine are shamelessly stolen from the TypeScript compiler, and it's used by millions of developers. The TypeScript team has fine-tuned these assumptions over many years to reach the best tradeoffs, maximizing the benefit while minimizing false-positive errors. For that reason, I'm very reluctant to tinker with those tradeoffs in Pyright.

**Comment by russelldavis** [comment]
There's one detail you may be overlooking that I think might change how you view this (continued below).

> There's no way for a type checker to know generally that a function call is going to mutate a non-local or global and assign it a different type.

In most languages (including typescript), that's true. But in python, a function cannot rebind a variable defined elsewhere without using a `global` or `nonlocal` declaration (I'll focus on `nonlocal` since that's where the issue is). So, if you see `nonlocal`, you actually *can* assume the function is going to rebind it (at least in some branch). Otherwise, the `nonlocal` statement shouldn't be there at all. Of course, it's possible to declare a `nonlocal` and then not rebind it, but it's clearly an error, and a user can't complain if the type checker makes assumptions based on it.

If you're still not convinced about the general case, let me at least convince you about a more specific problem. Let's look at a slightly modified version of the example. The function is now annotated as returning `int`, and `opt2` is returned after asserting it is not None:

```python
from typing import Optional
def bar() -> int:
    opt2: Optional[int] = None
    def qux():
        nonlocal opt2
        opt2 = 1
    mystery(qux)
    assert opt2 is not None
    return opt2 # Error: Type "Never" cannot be assigned to type "int"

def mystery(callback): ...
```

This is a well-typed program -- the error is a false positive.

BTW, there are legitmate use cases for this, which is why `nonlocal` was introduced in Python 3 -- see [PEP 3104](https://www.python.org/dev/peps/pep-3104/) for details. In my particular case, `mystery` is a third party asynchronous function that passes in its result to `qux`, which assigns that value to `opt2`. Synchronization primitives are used so that `bar` blocks after calling `mystery` until `qux` completes. Essentially, `bar` is a synchronous wrapper over the async `mystery`.

I'll also note that the equivalent version of this in typescript does *not* produce any type errors. [Check it out here](https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABAIwIYCcAUBKAXIsEAW2QFN1EBvAKETsQBtSpE4AHKAJn0JPMQA+BEAwaIAvMNG16TFgEcQADwmIcEgHxUZ9euy6qAjDsQBfE0QCeAZyjlLmRUuwmYwNfs4TxkwqOzaurpQABbocADuiABE4KRKbKTQpAAm0SbmuujMIOhIntTm1KCQsAiIVrb2mAD6EKiiaBAA1vgAYuDQ8GABlGbUQA). Typescript infers the type `never` just like pyright. But it allows the assignment from `never` to `int` to succeed. This makes sense to me -- `never` is the bottom type, so it should be assignable to everything. It seems like pyright should do the same.

If pyright did that, at least the error would go away. But I hope I've convinced you that it can do even better, given the semantics around `nonlocal`. For reference, mypy works the way I proposed.

**Comment by erictraut** [comment]
Are you suggesting that Pyright do global code flow analysis (i.e. code flow that spans multiple functions) for all possible modifications of a symbol along all code paths within a program? That's simply not feasible. It would be computationally very expensive and would be highly error prone. If there's any form of asynchrony in a program, knowing the order in which code is executed is impossible to determine statically. Short of doing global code flow analysis, I don't see how a type checker could possibly know that a subroutine call was going to mutate the value of one of its local variables.

Mypy doesn't appear to be doing anything resembling correct type tracking in this case, so I'm going to discount its behavior here. I simplified your code as follows. Note that I've added three `reveal_type` calls. I've also removed your `qux` function. So this should be a really straightforward case for any type checker.

```
def bar() -> int:
    opt2: Optional[int] = None
    reveal_type(opt2)
    mystery(lambda: None)
    reveal_type(opt2)
    assert opt2 is not None
    reveal_type(opt2)
    return opt2
```

Pyright correctly emits the following types:
* None
* None
* Never

Mypy emits the following, which makes no sense.
* Union[int, None]
* Union[int, None]
* int

If I now add back in your `qux` function and pass it in place of the empty lambda, both Pyright and mypy output exactly the same as above. So I don't think mypy is doing what you think it's doing. It's not doing anything special with your callback function or your nonlocal binding. It's simply exhibiting the same incorrect type tracking behavior that it does with the trivial case.

As for whether `Never` should be assignable to `int`, TypeScript defines its `never` type as being assignable to all types, so in that sense it's treating it like `any`. Pyright currently defines `Never` as assignable to no types. I can make a good case for either approach, but I'd rather stay consistent with TypeScript since I borrowed the idea from there, so I'm going to make that change. That will eliminate the error that you are seeing.

**Comment by russelldavis** [comment]
> Mypy emits the following, which makes no sense

Can you elaborate on why it doesn't make sense to you? Mypy is using the declared type for the first two, and for the last one, it's using type narrowing from the `assert` to remove `None` from the union.

Pyright will output the same thing if you force it to not narrow the initial type (change the first line to `opt2: Optional[int] = optint()`, where `optint` is a function that returns an `Optional[int]`).

---

You're right that the examples given don't show mypy doing anything special with `nonlocal`. When I mentioned mypy's behavior, I was playing around with a bunch of examples, and I didn't realize the ones I'd posted earlier wouldn't show it. Apologies.

The issue with those examples is that mypy's special behavior is to widen the type to include `int`, and in those examples, the type already includes `int` (in mypy only). Here's the example you're looking for, that shows the behavior:


```python
# Comments reflect types in mypy
def bar() -> None:
    opt2 = None
    reveal_type(opt2) # None

    def qux() -> None:
        nonlocal opt2
        opt2 = 1

    reveal_type(opt2) # Union[int, None]
    assert opt2 is not None
    reveal_type(opt2) # int
```

Note that mypy widens the type right after `qux` gets defined, so there's no call to `qux()` in this example. You'll see that if you comment out the definition of `qux`, or the line that declares opt2 as nonlocal, the types change. Note that with `qux` commented out, the third `reveal_type` won't result in any output, because mypy figures out that all code after the assert will be unreachable. This was quite confusing at first — pyright handles this much better, printing such types as `Never`.

> Are you suggesting that Pyright do global code flow analysis

Definitely not, and I agree it would be not be feasible. I don't know how mypy implemented this, but here's a simplified way to look at it: any time you see an assignment to a variable that has been declared as `nonlocal`, treat it exactly like you would as if the assignment were inside a conditional block in the outer function that the nonlocal refers to. So, the types for the example above would be treated the same as this:

```python
# Comments reflect types in both mypy and pyright
def bar() -> None:
    opt2 = None
    reveal_type(opt2) # None

    if random.choice([True, False]):
        opt2 = 1

    reveal_type(opt2) # Union[int, None]
    assert opt2 is not None
    reveal_type(opt2) # int
```

With this approach, there's no need for any global analysis. When you see `nonlocal foo` in a function `bar`, you just apply special handling to the assignments of `foo` directly in `bar`.

**Comment by erictraut** [comment]
Mypy's analysis doesn't make sense to me because it appears to be ignoring assignments in its analysis. When a symbol is assigned a value (such as `None` in your example), its type after that assignment statement should reflect that assignment. The declared type provides a starting point when no additional information is known, but assignments and conditional statements clearly affect the type. Mypy appears to be doing narrowing in some cases (for conditionals) but not in the simple assignment case. I've seen numerous cases where I disagree with mypy's type analysis. My goal isn't to replicate what its doing.

>any time you see an assignment to a variable that has been declared as nonlocal, treat it exactly as if the assignment

This would require global analysis in your example. Local analysis is restricted to a single function, such as `bar`. Global analysis refers to any analysis that requires the analyzer to consider the code within multiple functions (such as `bar` and `qux` in your example). Pyright's code flow analysis engine is local only. As it's analyzing `bar`, it doesn't even visit the code within `qux`.

Even if Pyright were to consider the code within `qux` as it was analyzing `bar`, it would be incorrect to make the assumption that you are suggesting. The location where `qux` is _declared_ is irrelevant for code flow analysis. What's important is where it is _invoked_. In your simplified example above, `qux` isn't invoked at all, so it should have no impact on the type analysis of `opt2`. Consider, for example, if the declaration of `qux` were moved up a few lines. Or if `qux` was outside of `bar`, and both were using a `nonlocal` binding to refer to some a variable that was declared in some common outer scope.

Like I said, I'm willing to change the behavior of assignment matching for `Never`.

The other changes you've suggested here are either incorrect or infeasible.

**Comment by russelldavis** [comment]
> Mypy's analysis doesn't make sense to me because it appears to be ignoring assignments in its analysis. When a symbol is assigned a value (such as None in your example), its type after that assignment statement should reflect that assignment.

Type narrowing in a case like that is a great feature, but I don't think it's generally considered necessary in order for types to make sense. Mypy's behavior there is pretty common — swift, kotlin, dart, sorbet, and pytype all do the same with the equivalent of that declaration.


> The location where qux is declared is irrelevant for code flow analysis.

Yep. Ideally the type widening would be applied the first time 'qux' is referenced, rather than when it is declared. I think mypy is doing the less ideal thing here in order to keep the analysis local (I can elaborate later).

That does result in types that are incorrect. But so does Pyright's approach — a variable shouldn't be `never` in a path that is reachable. If `never` is assignable to nothing, it can cause false positives. If `never` is assignable to everything, it can cause false negatives.

IMO, mypy's approach is less likely to cause problems in real world code (it didn't cause an issue with my code, and callbacks like mine are a common use case for `nonlocal`).

All that said, this is not that big an issue. I'll let it rest.

**Comment by erictraut** [comment]
The next version of Pyright will include the change where "Never" is assignable to all types (as opposed to not assignable to any types).

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.56, which I just published. It will also be in the next version of Pylance.
